Objects/listobject.c

```diff
@@ -1970,6 +2167,88 @@ listsort_impl(PyListObject *self, PyObject *keyfunc, int reverse)
         lo.values = saved_ob_item;
     }
 
+
+    /* The pre-sort check: here's where we decide which compare function to use.
+     * How much optimization is safe? We test for homogeneity with respect to 
+     * several properties that are expensive to check at compare-time, and 
+     * set ms appropriately. */
+    if (saved_ob_size > 1) {
+        /* Assume the first element is representative of the whole list. */
+        int keys_are_in_tuples = (lo.keys[0]->ob_type == &PyTuple_Type &&
```

Comment:
Actually, I just looked up the definition, and I don't think replacing the current code makes any sense:

```
#define PyTuple_CheckExact(op) (Py_TYPE(op) == &PyTuple_Type)
```